x-pack/plugins/session_view/common/types/process_tree/index.ts

```diff
@@ -0,0 +1,210 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+export enum EventKind {
+  event = 'event',
+  signal = 'signal',
+}
+
+export enum EventAction {
+  fork = 'fork',
+  exec = 'exec',
+  exit = 'exit',
+  output = 'output',
+}
+
+export interface EventActionPartition {
+  fork: ProcessEvent[];
+  exec: ProcessEvent[];
+  exit: ProcessEvent[];
+  output: ProcessEvent[];
+}
+
+export interface User {
+  id: string;
+  name: string;
+}
+
+export interface ProcessFields {
+  args: string[];
+  args_count: number;
+  command_line: string;
+  entity_id: string;
+  executable: string;
+  interactive: boolean;
+  name: string;
+  working_directory: string;
+  pid: number;
+  pgid: number;
+  user: User;
+  end?: string;
+  exit_code?: number;
+}
+
+export interface ProcessSelf extends ProcessFields {
+  parent: ProcessFields;
+  session: ProcessFields;
+  entry: ProcessFields;
+  last_user_entered?: ProcessFields;
+}
+
+export interface ProcessEventHost {
+  architecture: string;
+  hostname: string;
+  id: string;
+  ip: string;
+  mac: string;
+  name: string;
+  os: {
+    family: string;
+    full: string;
+    kernel: string;
+    name: string;
+    platform: string;
+    type: string;
+    version: string;
+  };
+}
+
+export interface ProcessEventAlertRule {
+  category: string;
+  consumer: string;
+  description: string;
+  enabled: boolean;
+  name: string;
+  query: string;
+  risk_score: number;
+  severity: string;
+  uuid: string;
+}
+
+export interface ProcessEventAlert {
+  uuid: string;
+  reason: string;
+  workflow_status: string;
+  status: string;
+  original_time: Date;
+  original_event: {
+    action: string;
+  };
+  rule: ProcessEventAlertRule;
+}
+
+export interface ProcessEvent {
+  '@timestamp': Date;
+  event: {
+    kind: EventKind;
+    category: string;
+    action: EventAction;
+  };
+  // optional host for now (raw agent output doesn't have server identity)
+  host?: ProcessEventHost;
+  process: ProcessSelf;
+  kibana?: {
+    alert: ProcessEventAlert;
+  };
+}
+
+export interface Process {
+  id: string; // the process entity_id
+  events: ProcessEvent[];
+  children: Process[];
+  parent: Process | undefined;
+  autoExpand: boolean;
+  searchMatched: string | null; // either false, or set to searchQuery
+  hasOutput(): boolean;
+  hasAlerts(): boolean;
+  getAlerts(): ProcessEvent[];
+  hasExec(): boolean;
+  getOutput(): string;
+  getDetails(): ProcessEvent;
+  isUserEntered(): boolean;
+  getMaxAlertLevel(): number | null;
+}
+
+export class ProcessImpl implements Process {
```

Comment:
Oh sorry I didn't really think about this on the first pass, but I guess this class really isn't a type per se...maybe we should find a better place to put this?

Reply:
What about converting `Process` to a class, and then `extends` it from `ProcessImpl`?

```js
class Process {
  id: string;
  events: ProcessEvent[];
  children: Process[];
  parent: Process | undefined;
  autoExpand: boolean;
  searchMatched: string | null;

  constructor(id: string) {
    this.id = id;
    this.events = [];
    this.children = [];
    this.autoExpand = false;
    this.searchMatched = null;
  }

  hasOutput() {
    // TODO: schema undecided
    return !!this.events.find(({ event }) => event.action === EventAction.output);
  }
 // ...other methods
}
```

implementation

```js
class ProcessImpl extends Process {
  constructor(id: string) {
    super(id);
  }
}
```

then our tests could mock like this instead?

```js
class SessionViewBasicProcessMock extends Process {
  constructor(id: string) {
    super('3f44d854-fe8d-5666-abc9-9efe7f970b4b');
    this.events = mockEvents;
  }
}
```

Reply:
Im of the opinion that we should only have types in type files. `Process` in your example is still a concrete class right? (I guess it is JS so theres really no true inheritance haha) 

are we actually going be to extending this? IMO we should avoid classes in general when working with react

Reply:
Yea you are right, it doesn't make sense to put it here. I am not sure where we should put it since we are thinking about refactoring process tree component, but I imagine after the refactor we are gonna move the `use_process_tree` hook to `public/components/ProcessTree`.  I thinking maybe we can just keep it in the `use_process_tree` for now? What do you all think?

Reply:
i think that makes sense 👍

Reply:
yeah, makes sense